cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java

```diff
@@ -0,0 +1,212 @@
+import co.cask.cdap.api.dataset.table.Put;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.internal.app.runtime.schedule.trigger.TriggerJsonCodec;
+import co.cask.cdap.internal.schedule.trigger.Trigger;
+import co.cask.cdap.proto.Notification;
+import co.cask.cdap.proto.id.ScheduleId;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Dataset that stores {@link Job}s, which correspond to schedules that have been triggered, but not yet executed.
+ *
+ * Row Key is in the following formats:
+ *   For Jobs:
+ *     'J':<partition_id>:<scheduleId>:<timestamp>
+ *     The <partition_id> is a hash based upon the scheduleId
+ *
+ *   For TMS MessageId:
+ *     'M':<topic>
+ */
+public class JobQueueDataset extends AbstractDataset implements JobQueue {
+
+  static final String EMBEDDED_TABLE_NAME = "t"; // table
+  private static final Gson GSON =
+    new GsonBuilder().registerTypeAdapter(Trigger.class, new TriggerJsonCodec()).create();
+
+  // For now, simply serialize the entire Job into one column
+  private static final byte[] COL = new byte[] {'C'};
+  private static final byte[] JOB_ROW_PREFIX = new byte[] {'J'};
+  private static final byte[] ROW_KEY_SEPARATOR = new byte[] {':'};
+  private static final byte[] MESSAGE_ID_ROW_PREFIX = new byte[] {'M'};
+
+  private static final int NUM_PARTITIONS = 16;
+
+  private final Table table;
+
+  JobQueueDataset(String instanceName, @EmbeddedDataset(EMBEDDED_TABLE_NAME) Table table) {
+    super(instanceName, table);
+    this.table = table;
+  }
+
+  @Override
+  public CloseableIterator<Job> getJobsForSchedule(ScheduleId scheduleId) {
+    byte[] keyPrefix = getRowKeyPrefix(scheduleId);
+    return createCloseableIterator(table.scan(keyPrefix, Bytes.stopKeyForPrefix(keyPrefix)));
+  }
+
+  @Override
+  public Job getJob(JobKey jobKey) {
+    Row row = table.get(getRowKey(jobKey.getScheduleId(), jobKey.getCreationTime()));
+    if (row.isEmpty()) {
+      return null;
+    }
+    return fromRow(row);
+  }
+
+  @Override
+  public void put(Job job) {
+    table.put(toPut(job));
+  }
+
+  @Override
+  public void updateState(Job job, Job.State state) {
+    // assert that the job state transition is valid
+    job.getState().check(state);
+    put(new SimpleJob(job.getSchedule(), job.getCreationTime(), job.getNotifications(), state));
+  }
+
+  @Override
+  public void updateNotification(Job job, Notification notification) {
+    List<Notification> notifications = new ArrayList<>(job.getNotifications());
+    notifications.add(notification);
+    put(new SimpleJob(job.getSchedule(), job.getCreationTime(), notifications, job.getState()));
```

Comment:
why isn't there a method in `Job` to allow appending to its notifications list directly?